sled-agent/zone-images/src/mupdate_override.rs

```diff
@@ -0,0 +1,1006 @@
+            .all_m2_zpools
+            .iter()
+            .filter(|&zpool_name| zpool_name != boot_zpool);
+        let non_boot_disks_overrides = non_boot_zpools
+            .map(|zpool_name| {
+                let dataset =
+                    zpool_name.dataset_mountpoint(zpools.root, INSTALL_DATASET);
+
+                let (path, res) = read_mupdate_override(log, &dataset);
+                MupdateOverrideNonBootInfo {
+                    zpool_name: *zpool_name,
+                    path,
+                    result: MupdateOverrideNonBootResult::new(
+                        res,
+                        &boot_disk_res,
+                    ),
+                }
+            })
+            .collect();
+
+        Self {
+            boot_zpool: *boot_zpool,
+            boot_disk_path,
+            boot_disk_override: boot_disk_res,
+            non_boot_disk_overrides: non_boot_disks_overrides,
+        }
+    }
+
+    pub(crate) fn log_results(&self, log: &slog::Logger) {
+        let log = log.new(o!(
+            "boot_zpool" => self.boot_zpool.to_string(),
+            "boot_disk_path" => self.boot_disk_path.to_string(),
+        ));
+
+        match &self.boot_disk_override {
+            Ok(Some(mupdate_override)) => {
+                info!(
+                    log,
+                    "found mupdate override for boot disk";
+                    "data" => ?mupdate_override,
+                );
+            }
+            Ok(None) => {
+                info!(log, "no mupdate override for boot disk");
+            }
+            Err(error) => {
+                // This error most likely requires operator intervention -- if
+                // it happens, we'll continue to bring sled-agent up but reject
+                // all zone image lookups.
+                error!(
+                    log,
+                    "error reading mupdate override for boot disk, \
+                     will not bring up zones";
+                    "error" => InlineErrorChain::new(error),
+                );
+            }
+        }
+
+        if self.non_boot_disk_overrides.is_empty() {
+            warn!(
+                log,
+                "no non-boot zpools found, unable to verify consistency -- \
+                 this may be a hardware issue with the non-boot M.2"
+            );
+        }
+
+        for info in &self.non_boot_disk_overrides {
+            info.log_result(&log);
+        }
+    }
+}
+
+fn read_mupdate_override(
+    log: &slog::Logger,
+    dataset_dir: &Utf8Path,
+) -> (Utf8PathBuf, Result<Option<MupdateOverrideInfo>, MupdateOverrideReadError>)
+{
+    let override_path = dataset_dir.join(MupdateOverrideInfo::FILE_NAME);
+
+    fn inner(
+        log: &slog::Logger,
+        dataset_dir: &Utf8Path,
+        override_path: &Utf8Path,
+    ) -> Result<Option<MupdateOverrideInfo>, MupdateOverrideReadError> {
+        // First check that the dataset directory exists. It would be nice if we
```

Comment:
I was going to ask why we want to check the directory exists instead of just trying to read the file. I think this is to distinguish "the install dataset isn't here at all" (an error) from "the install dataset exists but doesn't have the override file" (not an error)? If that's right, it might be worth noting in this comment.

Reply:
Expanded the comment.